### PR TITLE
Adjust "hack/make/.detect-daemon-osarch" to be the source of truth for "platform detection"

### LINKED ISSUE #21125: hack/make/.detect-daemon-osarch fails on linux/arm
**Output of `docker version`:**
```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.3.3
Git commit (client): 7c8fca2
OS/Arch (client): linux/arm
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.3.3
Git commit (server): 7c8fca2
OS/Arch (server): linux/arm
```

**Output of `docker info`:**
```
Storage Driver: btrfs
 Build Version: Btrfs v3.17
 Library Version: 101
Execution Driver: native-0.2
Kernel Version: 4.4.1-sunxi
Operating System: Debian GNU/Linux 8 (jessie)
```

When trying to build the docker-dev image (with `make build`) I noticed that the script was using the usual Dockerfile instead of the Dockerfile.armhf, which of course resulted in a failure. (Can't run the AMD64 image on armhf)
I looked into the Makefile and saw that it is pulling the "DOCKER_OSARCH" variable from the script hack/make/.detect-daemon-osarch, which exports DOCKER_ENGINE_OSARCH, DOCKER_ENGINE_GOOS, DOCKER_ENGINE_GOARCH and DOCKER_CLIENT_OSARCH

**Describe the results you received:**
```
/bin/bash
$ source hack/make/.detect-daemon-osarch
$ echo $DOCKER_ENGINE_OSARCH
(empty line)
$ echo $DOCKER_CLIENT_OSARCH
(empty line)
```
The variables DOCKER_ENGINE_GOOS and DOCKER_ENGINE_GOARCH both depend on DOCKER_ENGINE_OSARCH, therefore they are false, too.

**Describe the results you expected:**
```
/bin/bash
$ source hack/make/.detect-daemon-osarch
$ echo $DOCKER_ENGINE_OSARCH
linux/arm
$ echo $DOCKER_CLIENT_OSARCH
linux/arm
```

**Additional information you deem important (e.g. issue happens only occasionally):**
```
$ uname -srmo
Linux 4.4.1-sunxi armv7l GNU/Linux
$ cat /etc/debian_version
8.3
```

### PR DESCRIPTION
**Pull request by tianon**
Instead of being split between three files, let's let `hack/make/.detect-daemon-osarch` be our single source of truth for multiarch detection/vars.  Not only does it make it slightly easier to make sure we change everything properly when these bits have to change, but it also makes it so that all bits of `hack/make.sh` (especially `hack/make/.ensure-frozen-images`) work properly outside the context of the `Makefile` on all platforms.

Fixes #21125

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,30 +1,8 @@
...
-	DOCKERFILE := Dockerfile.aarch64
-else
-ifeq ($(DOCKER_OSARCH), linux/ppc64le)
-	DOCKERFILE := Dockerfile.ppc64le
-else
-ifeq ($(DOCKER_OSARCH), linux/s390x)
-	DOCKERFILE := Dockerfile.s390x
-else
-ifeq ($(DOCKER_OSARCH), windows/amd64)
-	DOCKERFILE := Dockerfile.windows
```
Comment (tianon) [inline Makefile]: #ohman   need to do something here too; will update shortly.

Comment (tianon) [inline Makefile]: handled now :+1:

**Comment by thaJeztah** [comment]
WindowsTP4 flaky tests; https://github.com/docker/docker/issues/20819 (https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/2476/console)

```
05:37:02 FAIL: docker_cli_ps_test.go:475: DockerSuite.TestPsListContainersFilterExited
05:37:02 
05:37:02 docker_cli_ps_test.go:500:
05:37:02     c.Assert(ids, checker.HasLen, 2, check.Commentf("Should be 2 zero exited containers got %d: %s", len(ids), out))
05:37:02 ... obtained []string = []string{"c93532cbfef6d6234c45be925ddd600af0b859b599d6b417dbe5b28c6d740b03", "ed44e9176ef4486204278c8bb7bbcb76a46aaaa56adfaf0ff77903bc43db5560", "7c5ca77935b4e617eddb67c7ddbefb1a489d9e5d7f214b1c3e2ed0db0659ca2a"}
05:37:02 ... n int = 2
05:37:02 ... Should be 2 zero exited containers got 3: c93532cbfef6d6234c45be925ddd600af0b859b599d6b417dbe5b28c6d740b03
05:37:02 ed44e9176ef4486204278c8bb7bbcb76a46aaaa56adfaf0ff77903bc43db5560
05:37:02 7c5ca77935b4e617eddb67c7ddbefb1a489d9e5d7f214b1c3e2ed0db0659ca2a
05:37:02 
```

**Comment by tianon** [comment]
I've added a second small commit here which will fix #21125 too. :smile:

File: Makefile [hunk 2]
```diff
@@ -37,7 +15,6 @@ DOCKER_ENVS := \
 	-e DOCKER_CLIENTONLY \
 	-e DOCKER_DEBUG \
 	-e DOCKER_EXPERIMENTAL \
-	-e DOCKERFILE \
```
Comment (clnperez) [inline Makefile]: Will this omission make it impossible to use the gccgo Dockerfile with make.sh?

Comment (tianon) [inline Makefile]: It should still be overwritable in the same way via something like `make DOCKERFILE=Dockerfile.gccgo gccgo` just like before -- this variable being passed in to the container (which is what this line was doing) was controlling where the "frozen images" were being pulled from if they weren't available in the image already, which likely would've been broken with a different `DOCKERFILE=...` anyhow.

Comment (tianon) [inline Makefile]: (Where this way, we'll choose which `Dockerfile` to steal that line from as-necessary dynamically based on our platform regardless of which `Dockerfile` we're building from.)

**Comment by thaJeztah** [comment]
ping @clnperez does this LGTY?

**Comment by clnperez** [comment]
@thaJeztah -- Yes. Thanks for the clarification @tianon

**Comment by calavera** [comment]
LGTM

**Comment by estesp** [comment]
LGTM

**Pull request merged**